Ray 2.5.0 breaks mypy

### What happened + What you expected to happen

Run mypy on code imports `ray` functions throws new errors which doesn't exists on 2.4.0

```
error: Module "ray" does not explicitly export attribute "wait"  [attr-defined]
error: Module "ray" does not explicitly export attribute "init"  [attr-defined]
...
```

This is caused by #34730, `__all__` should be simple literals, not a complex computed value, mypy only has very limited support for computed `__all__` values.

### Versions / Dependencies

ray 2.5.0
mypy 1.3.0

### Reproduction script

```
import ray

@ray.remote
def foo() -> None:
    pass
```

Save as `foo.py` and run `mypy foo.py`

### Issue Severity

High: It blocks me from completing my task.

@raulchen I was using Python 3.10.6 on Ubuntu 22.04.

Here is the same repro on macOS with Python 3.10 with a `strict = True` in `mypy.ini`. I think the mypy configuration is causing the difference:

```bash
$ cd /tmp
$ python3.10 -m venv venv
$ . venv/bin/activate
$ cat 1.py
import ray

@ray.remote
def foo() -> None:
    pass

$ pip install ray mypy

$ ray --version
ray, version 2.5.0
$ mypy --version
mypy 1.3.0 (compiled: yes)

$ cat mypy.ini
[mypy]
strict = True

$ mypy 1.py
1.py:3: error: Module "ray" does not explicitly export attribute "remote"  [attr-defined]
1.py:3: error: Untyped decorator makes function "foo" untyped  [misc]
Found 2 errors in 1 file (checked 1 source file)
```

--- a/python/ray/__init__.py
+++ b/python/ray/__init__.py
@@ -183,7 +183,8 @@ serialization = _DeprecationWrapper("serialization", ray._private.serialization)
 state = _DeprecationWrapper("state", ray._private.state)
 
 
-RAY_APIS = {
+# Pulic Ray APIs
+__all__ = [
     "__version__",
     "_config",
     "get_runtime_context",
@@ -214,7 +215,7 @@ RAY_APIS = {
     "LOCAL_MODE",
     "SCRIPT_MODE",
     "WORKER_MODE",
-}
+]
 
 # Public APIs that should automatically trigger ray.init().
 AUTO_INIT_APIS = {
@@ -254,14 +255,11 @@ NON_AUTO_INIT_APIS = {
     "timeline",
 }
 
-assert RAY_APIS == AUTO_INIT_APIS | NON_AUTO_INIT_APIS
+assert set(__all__) == AUTO_INIT_APIS | NON_AUTO_INIT_APIS
 from ray._private.auto_init_hook import wrap_auto_init_for_all_apis  # noqa: E402
 
 wrap_auto_init_for_all_apis(AUTO_INIT_APIS)
 del wrap_auto_init_for_all_apis
-
-
-__all__ = list(RAY_APIS)
 
 # Subpackages
 __all__ += [

